report/topology.go

```diff
@@ -195,6 +197,100 @@ func (c Counters) Copy() Counters {
 	return result
 }
 
+// Sets is a string->set-of-strings map.
+type Sets map[string]StringSet
+
+// Merge merges two sets maps into a fresh set, performing set-union merges as
+// appropriate.
+func (s Sets) Merge(other Sets) Sets {
+	result := s.Copy()
+	for k, v := range other {
+		result[k].Merge(v)
+	}
+	return result
+}
+
+// Copy returns a value copy of the sets map.
+func (s Sets) Copy() Sets {
+	result := Sets{}
+	for k, v := range s {
+		result[k] = v.Copy()
+	}
+	return result
+}
+
+// StringSet is a sorted set of unique strings. Clients must use the Add
+// method to add strings.
+type StringSet []string
```

Comment:
The implementation is contained to this file, and well-tested. The public API is Make, Add, Merge, and Contains. Those methods would remain with the other implementation.

Also, another important benefit of []string is that the JSON serialization is intuitive: `["a", "b", "c"]`. If we switched to map[string]struct{}, we would need to implement Marshal/Unmarshal JSON methods to serialize to, and deserialize from, this form.